### PR TITLE
[go-sqlcipher] replace AES implementation with mbedtls

### PR DESCRIPTION
**Pull request by caybro**
Update `status-im/go-sqlcipher` to `v0.1.0-status.6`

HW accelerated versions provided for x86_64 and arm64, with compile time
guards and runtime detection

needed for https://github.com/status-im/status-desktop/issues/10726

### CODE REVIEW

**Comment by caybro** [comment]
With this change, we finally have a sane CPU usage in the desktop app, without any glaring bottlenecks:

Snímek obrazovky z 2023-05-23 01-07-28

**Comment by caybro** [comment]
The tests just keep on failing (timing out) after ~20 minutes :/

**Comment by jrainville** [comment]
> The tests just keep on failing (timing out) after ~20 minutes :/

Yeah, they do that sometimes. You have to retry some more :confused:

**Review by jrainville — APPROVED** [review]

File: vendor/github.com/mutecomm/go-sqlcipher/aes.c [hunk 1]
```diff
@@ -284,172 +123,26 @@ int SETUP(const unsigned char *key, int keylen, int num_rounds, symmetric_key *s
...
-    return CRYPT_OK;
+    return mbedtls_aes_crypt_ecb(&ctx_encrypt, MBEDTLS_AES_ENCRYPT, pt, ct) == 0 ? CRYPT_OK : CRYPT_ERROR;
 }
 
 #ifdef LTC_CLEAN_STACK
-int ECB_ENC(const unsigned char *pt, unsigned char *ct, symmetric_key *skey) 
+int ECB_ENC(const unsigned char *pt, unsigned char *ct, symmetric_key *skey)
 {
-   int err = _rijndael_ecb_encrypt(pt, ct, skey);
-   burn_stack(sizeof(unsigned long)*8 + sizeof(unsigned long*) + sizeof(int)*2);
```
Comment (alexjba) [inline vendor/github.com/mutecomm/go-sqlcipher/aes.c:451]: Missing burn_stack ?

File: vendor/github.com/mutecomm/go-sqlcipher/aes.c [hunk 2]
```diff
@@ -463,173 +156,27 @@ int ECB_ENC(const unsigned char *pt, unsigned char *ct, symmetric_key *skey)
...
+    return mbedtls_aes_crypt_ecb(&ctx_decrypt, MBEDTLS_AES_DECRYPT, ct, pt) == 0 ? CRYPT_OK : CRYPT_ERROR;
 }
 
 
 #ifdef LTC_CLEAN_STACK
-int ECB_DEC(const unsigned char *ct, unsigned char *pt, symmetric_key *skey) 
+int ECB_DEC(const unsigned char *ct, unsigned char *pt, symmetric_key *skey)
 {
-   int err = _rijndael_ecb_decrypt(ct, pt, skey);
-   burn_stack(sizeof(unsigned long)*8 + sizeof(unsigned long*) + sizeof(int)*2);
```
Comment (alexjba) [inline vendor/github.com/mutecomm/go-sqlcipher/aes.c:631]: Same here, missing burn_stack?

File: vendor/github.com/mutecomm/go-sqlcipher/aes.c [hunk 3]
```diff
@@ -726,6 +273,10 @@ int ECB_TEST(void)
 */
 void ECB_DONE(symmetric_key *skey)
 {
+    mbedtls_aes_free(&ctx_encrypt);
```
Comment (alexjba) [inline vendor/github.com/mutecomm/go-sqlcipher/aes.c:276]: Is it on purpose to do it twice if `ENCRYPT_ONLY` is not defined?

**Review by osmaczko — APPROVED** [review]
:muscle:

**Comment by caybro** [comment]
> > The tests just keep on failing (timing out) after ~20 minutes :/
> 
> Yeah, they do that sometimes. You have to retry some more confused

Restarted at least a dozen times... to no avail :(

File: vendor/github.com/mutecomm/go-sqlcipher/aes.c (hunk 3 above)
Comment (caybro) [reply vendor/github.com/mutecomm/go-sqlcipher/aes.c:276]: Yes, different contexts (encrypt + decrypt)

File: vendor/github.com/mutecomm/go-sqlcipher/aes.c (hunk 1 above)
Comment (caybro) [reply vendor/github.com/mutecomm/go-sqlcipher/aes.c:451]: Line 578:

```c
mbedtls_platform_zeroize(&t, sizeof(t));
```

It just does it at a different place, but still there :)

File: vendor/github.com/mutecomm/go-sqlcipher/aes.c (hunk 2 above)
Comment (caybro) [reply vendor/github.com/mutecomm/go-sqlcipher/aes.c:631]: See above

**Review by alexjba — APPROVED** [review]
Nice! 👍

**Comment by caybro** [comment]
The tests seem to consistently fail with:
```
[2023-05-24T10:31:34.201Z] ok  	github.com/status-im/status-go/wakuv2/common	(cached)
[2023-05-24T10:31:34.201Z] ?   	github.com/status-im/status-go/wakuv2/persistence	[no test files]
[2023-05-24T10:31:34.201Z] FAIL
[2023-05-24T10:31:34.202Z] make: *** [Makefile:314: test-unit] Error 1
```

Any idea?

**Comment by osmaczko** [comment]
> The tests seem to consistently fail with:
> 
> ```
> [2023-05-24T10:31:34.201Z] ok  	github.com/status-im/status-go/wakuv2/common	(cached)
> [2023-05-24T10:31:34.201Z] ?   	github.com/status-im/status-go/wakuv2/persistence	[no test files]
> [2023-05-24T10:31:34.201Z] FAIL
> [2023-05-24T10:31:34.202Z] make: *** [Makefile:314: test-unit] Error 1
> ```
> 
> Any idea?

I think it fails much earlier:

```
[2023-05-24T12:48:23.320Z] === RUN   TestEncryptionServiceTestSuite/TestRandomised
[2023-05-24T12:48:23.321Z] Starting test with seed: 176214e640e47ce4
[2023-05-24T12:48:23.321Z] fatal error: unexpected signal during runtime execution
[2023-05-24T12:48:23.321Z] [signal SIGSEGV: segmentation violation code=0x1 addr=0x110f000 pc=0x979cd6]
...
...
...
[2023-05-24T12:48:23.425Z] FAIL github.com/status-im/status-go/protocol/encryption      826.680s
```

**Comment by osmaczko** [comment]
> ```
> [2023-05-24T12:48:23.320Z] === RUN   TestEncryptionServiceTestSuite/TestRandomised
> [2023-05-24T12:48:23.321Z] Starting test with seed: 176214e640e47ce4
> [2023-05-24T12:48:23.321Z] fatal error: unexpected signal during runtime execution
> [2023-05-24T12:48:23.321Z] [signal SIGSEGV: segmentation violation code=0x1 addr=0x110f000 pc=0x979cd6]
> ```
Callstack:
```
runtime.throw (/usr/local/go/src/runtime/panic.go:1040)
runtime.sigpanic (/usr/local/go/src/runtime/signal_unix.go:821)
C.mbedtls_aesni_crypt_ecb (/home/osmaczny/Dev/status/status-go/vendor/github.com/mutecomm/go-sqlcipher/aesni.c:85)
C.mbedtls_aes_crypt_ecb (/home/osmaczny/Dev/status/status-go/vendor/github.com/mutecomm/go-sqlcipher/mbtls_aes.c:699)
C._rijndael_ecb_encrypt (/home/osmaczny/Dev/status/status-go/vendor/github.com/mutecomm/go-sqlcipher/aes.c:139)
C.rijndael_ecb_encrypt (/home/osmaczny/Dev/status/status-go/vendor/github.com/mutecomm/go-sqlcipher/aes.c:145)
C.cbc_encrypt (/home/osmaczny/Dev/status/status-go/vendor/github.com/mutecomm/go-sqlcipher/cbc_encrypt.c:71)
C.sqlcipher_ltc_cipher (/home/osmaczny/Dev/status/status-go/vendor/github.com/mutecomm/go-sqlcipher/sqlite3.c:21510)
C.sqlcipher_page_cipher (/home/osmaczny/Dev/status/status-go/vendor/github.com/mutecomm/go-sqlcipher/sqlite3.c:20909)
C.sqlite3Codec (/home/osmaczny/Dev/status/status-go/vendor/github.com/mutecomm/go-sqlcipher/sqlite3.c:19459)
C.pagerAddPageToRollbackJournal (/home/osmaczny/Dev/status/status-go/vendor/github.com/mutecomm/go-sqlcipher/sqlite3.c:57232)
C.pager_write (/home/osmaczny/Dev/status/status-go/vendor/github.com/mutecomm/go-sqlcipher/sqlite3.c:57322)
C.sqlite3PagerWrite (/home/osmaczny/Dev/status/status-go/vendor/github.com/mutecomm/go-sqlcipher/sqlite3.c:57476)
C.pager_incr_changecounter (/home/osmaczny/Dev/status/status-go/vendor/github.com/mutecomm/go-sqlcipher/sqlite3.c:57586)
C.sqlite3PagerCommitPhaseOne (/home/osmaczny/Dev/status/status-go/vendor/github.com/mutecomm/go-sqlcipher/sqlite3.c:57797)
C.sqlite3BtreeCommitPhaseOne (/home/osmaczny/Dev/status/status-go/vendor/github.com/mutecomm/go-sqlcipher/sqlite3.c:66672)
C.vdbeCommit (/home/osmaczny/Dev/status/status-go/vendor/github.com/mutecomm/go-sqlcipher/sqlite3.c:77588)
C.sqlite3VdbeHalt (/home/osmaczny/Dev/status/status-go/vendor/github.com/mutecomm/go-sqlcipher/sqlite3.c:77989)
C.sqlite3VdbeExec (/home/osmaczny/Dev/status/status-go/vendor/github.com/mutecomm/go-sqlcipher/sqlite3.c:83244)
C.sqlite3Step (/home/osmaczny/Dev/status/status-go/vendor/github.com/mutecomm/go-sqlcipher/sqlite3.c:80626)
```

**Comment by caybro** [comment]
> > ```
> > [2023-05-24T12:48:23.320Z] === RUN   TestEncryptionServiceTestSuite/TestRandomised
> > [2023-05-24T12:48:23.321Z] Starting test with seed: 176214e640e47ce4
> > [2023-05-24T12:48:23.321Z] fatal error: unexpected signal during runtime execution
> > [2023-05-24T12:48:23.321Z] [signal SIGSEGV: segmentation violation code=0x1 addr=0x110f000 pc=0x979cd6]
> > ```
> 
> Callstack:
> 
> ```
> runtime.throw (/usr/local/go/src/runtime/panic.go:1040)
> runtime.sigpanic (/usr/local/go/src/runtime/signal_unix.go:821)
> C.mbedtls_aesni_crypt_ecb (/home/osmaczny/Dev/status/status-go/vendor/github.com/mutecomm/go-sqlcipher/aesni.c:85)

@osmaczko yup, I got the same crash; it also corrupts the DB. The bad news is that it started happening after I rebased this PR on top of `develop`, that is the `DB pool` stuff. It works fine without :( I locally reverted 5777bb429a196672f21db97e5f3251ed674c6280 and it works fine:

App in action, discord import going on:
image

Hotspot session capturing the same:
image

No crypto functions (AES or SHA1) in sight, app stays smooth

**Comment by caybro** [comment]
> caybro — today at 17:29
> lol, the discord import already started meanwhile and boy it's fast now 

> 2023-05-24T17:44:18.340+0200	INFO	communities/manager.go:254	saving 3/3 chunk with 139 app messages
> 2023-05-24T17:44:18.413+0200	INFO	communities/manager.go:254	no message archives to import

The whole Discord import took 15 minutes in total, wow!

**Comment by osmaczko** [comment]
> @osmaczko yup, I got the same crash; it also corrupts the DB. The bad news is that it started happening after I rebased this PR on top of `develop`, that is the `DB pool` stuff. It works fine without :( I locally reverted [5777bb4](https://github.com/status-im/status-go/commit/5777bb429a196672f21db97e5f3251ed674c6280) and it works fine:

This leads me to suspect that the mbedtls AES implementation may not be thread-safe or properly handle concurrent access.

File: vendor/github.com/mutecomm/go-sqlcipher/aes.c [hunk 4]
```diff
@@ -34,7 +18,12 @@
 
 #ifdef LTC_RIJNDAEL
 
-#ifndef ENCRYPT_ONLY 
+#include "mbtls_aes.h"
+
+static mbedtls_aes_context ctx_encrypt;
+
+#ifndef ENCRYPT_ONLY
+static mbedtls_aes_context ctx_decrypt;
```
Comment (alexjba) [inline vendor/github.com/mutecomm/go-sqlcipher/aes.c]: Using thread local variables seems to fix the DB corruption (hopefully the x86_64 crash also)

```suggestion
static __thread mbedtls_aes_context ctx_encrypt;

#ifndef ENCRYPT_ONLY
static __thread mbedtls_aes_context ctx_decrypt;
```

**Review by alexjba — CHANGES_REQUESTED** [review]

File: vendor/github.com/mutecomm/go-sqlcipher/aes.c (hunk 4 above)
Comment (caybro) [reply vendor/github.com/mutecomm/go-sqlcipher/aes.c]: Will give it a try!

Comment (caybro) [reply vendor/github.com/mutecomm/go-sqlcipher/aes.c]: works \o/

**Review by alexjba — APPROVED** [review]
🚀

**Pull request merged**